File: 002_credit_card/credit.py
from cs50 import get_int

# Get the user's card number.
number = get_int("Please input your card number here: ")

digits_total = 0
digit_count = len(str(number))

# Add up every second digit starting from the end of the number.
for i in range(digit_count, 0, -2):
    digits_total += int(str(number)[i - 1])

# Double every second digit starting from the second number at the end of the number.
for i in range(digit_count - 1, 0, -2):
    single_digit = int(str(number)[i - 1]) * 2

    # In case of double digits after multiplication, seperate the digits and add to sum.
    if single_digit > 9:
        digits_total += int(str(single_digit)[0]) + int(str(single_digit)[1])
    else:
        digits_total += single_digit

# Check if the checksum has a remainder of 0.
if (digits_total % 10 == 0):
    # Get the first two digits of the card number.
    first_digits = int(str(number)[:2])

    # Determine the card type or if invalid based on the first two card digits.
    if digit_count == 13 and first_digits >= 40 and first_digits <= 49:
        print("VISA")
    elif digit_count == 15 and (first_digits == 34 or first_digits == 37):
        print("AMEX")
    elif digit_count == 16 and first_digits >= 40 and first_digits <= 49:
        print("VISA")
    elif digit_count == 16 and first_digits >= 51 and first_digits <= 55:
        print("MASTERCARD")
    else:
        print("INVALID")
else:
    print("INVALID")

File: 002_credit_card/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    long number = get_long("Please input your card number here: ");
    long number_backup = number;
    int digit_counter = 0;
    int second_digit = 0;
    int remaining_digit = 0;
    int i;

    // determine the length of the number
    for (i = 0; number_backup > 0; i++)
    {
        if (i % 2 == 0)
        {
            // get and add up every second digit starting from the end of the number
            int single_digit = number_backup % 10;
            number_backup /= 10;
            remaining_digit += single_digit;
        }
        else
        {
            // get every second digit starting from the second number at the end of the number
            int single_digit = number_backup % 10 * 2;
            number_backup /= 10;

            // in case of double digits after multiplication, seperate the digits and add to sum
            if (single_digit > 9)
            {
                second_digit += single_digit % 10;
                single_digit /= 10;
                second_digit += single_digit;
            }
            else
            {
                // add the digit to the sum
                second_digit += single_digit;
            }
        }
        digit_counter++;
    }

    // calculate the checksum
    int checksum = second_digit + remaining_digit;

    // check the card number length and if the checksum has a remainder of 0
    if ((checksum % 10 == 0) && (digit_counter == 13 || digit_counter == 15 || digit_counter == 16))
    {
        // get the first two digits of the card number
        float first_digits = floor(number / pow(10, (digit_counter - 2)));

        // determine the card type and validity based on the number length and first two card digits
        switch (digit_counter)
        {
            case 13:
                if (first_digits >= 40 && first_digits <= 49)
                {
                    printf("VISA\n");
                    break;
                }
            case 15:
                if (first_digits == 34 || first_digits == 37)
                {
                    printf("AMEX\n");
                    break;
                }
            case 16:
                if (first_digits >= 40 && first_digits <= 49)
                {
                    printf("VISA\n");
                    break;
                }
                else if (first_digits >= 51 && first_digits <= 55)
                {
                    printf("MASTERCARD\n");
                    break;
                }
            default:
                printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}
